Handle missing objects gracefully

alizenguyen opened the pull request:
<!--- Note: Please open the PR in draft form until you are ready for active review. -->
Closes #4258 <!--- Insert Issue Number(s) this PR addresses. Start by typing # will open a dropdown of recent issues. Note: this does not work on PRs which target release branches -->

### Describe your changes:
<!--- Describe your changes and add any comments about your approach either here or inline if code comments aren't added -->

Added checks for missing data as a result of missing domain objects to fix console errors located in a display layouts, overlay plots, and stacked plots as shown in the issue #4258.

The result should only display console warnings of missing objects. Stacked plots should have an empty space with no errors shown. See image below.

<img width="1741" alt="Screen Shot 2022-06-27 at 9 18 38 PM" src="https://user-images.githubusercontent.com/29556173/176077444-892033d7-1a5f-47a0-a46f-7ed869e3203d.png">

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [x] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

unlikelyzero commented on src/api/telemetry/TelemetryAPI.js:
```diff
@@ -529,7 +529,7 @@ define([
 
         this.openmct.notifications.error(message);
-        console.error(detailMessage);
+        console.warn(detailMessage);
```
is there a scenario where this should still throw an error instead of a warn?

alizenguyen replied:
Hmm I'm not sure... @jvigliotta any incite? It does send an error notification as well.

unlikelyzero replied:
@alizenguyen let's add an e2e test for this

jvigliotta replied:
I think the error notification is sufficient. I think there are many times where a provider doesn't exist for an object, I don't want to throw console errors for that each time, I think a warning would suffice. I mean it's evident from the view as well...

unlikelyzero commented on e2e/tests/plugins/plot/missingPlotObj.e2e.spec.js:
```diff
@@ -0,0 +1,134 @@
+ *****************************************************************************/
+
+/*
+Tests to verify log plot functionality. Note this test suite if very much under active development and should not
```
lets update this text

unlikelyzero commented on e2e/tests/plugins/plot/missingPlotObj.e2e.spec.js:
```diff
@@ -0,0 +1,134 @@
+
+        //Sets local storage with missing object
+        await page.evaluate(
+            `window.localStorage.setItem('mct', '${JSON.stringify(parsedData)}')`
```
![whoa](https://user-images.githubusercontent.com/12446869/176764722-f5990980-f203-473d-a768-af0d23ec2731.gif)

unlikelyzero commented on e2e/tests/plugins/plot/missingPlotObj.e2e.spec.js:
```diff
@@ -0,0 +1,134 @@
+        );
+
+        //Reloads page and clicks on stacked plot
+        page.reload();
```
note, reloads on vanilla openmct are known to be troublesome. You should consider putting some sort of logic after  line 50 to verify a properly "settled" DOM

unlikelyzero commented on e2e/tests/plugins/plot/missingPlotObj.e2e.spec.js:
```diff
@@ -0,0 +1,134 @@
+    });
+});
+
+async function makeStackedPlot(page) {
```
can you add some type hint intellisense here? If they're exclusively used in the context of this suite you can also give it `@private`

unlikelyzero commented on e2e/tests/plugins/plot/missingPlotObj.e2e.spec.js:
```diff
@@ -0,0 +1,134 @@
+        ]);
+
+        //Check that there is only one stacked item plot with a plot, the missing one will be empty
+        await expect(page.locator(".c-plot--stacked-container:has(.gl-plot)")).toHaveCount(1);
```
you should be able to intercept page.console errors and verify that console.warn is thrown with this API https://playwright.dev/docs/api/class-consolemessage

unlikelyzero reviewed (approved):
This new suite looks amazing. I'm putting this into 'approve' so that this PR doesn't get locked while i'm OOO tomorrow.

There are a few comments which _should_ be addressed but are not 100% necessary

alizenguyen commented:
@unlikelyzero I think I addressed all your comments. It dismissed your approval when I pushed the changes. I'll see if someone can review it tomorrow. Enjoy your day off!

jvigliotta commented on src/api/telemetry/TelemetryCollection.js:
```diff
@@ -322,7 +322,7 @@ export class TelemetryCollection extends EventEmitter {
      */
     _setTimeSystem(timeSystem) {
-        let domains = this.metadata.valuesForHints(['domain']);
+        let domains = this.metadata ? this.metadata.valuesForHints(['domain']) : [];
```
Can we set domain to it's default of `[]` right off the bat? Then can we just get one check for this.metada and if it exists, then we can set domain to that (as well as metadatavalue as it also relies on this.metadata).

Something like:
``` js
let domains = [];
let metadataValue = { format: timeSystem.key };

if (this.metadata) { ... update if necessary }
```

jvigliotta commented on src/plugins/displayLayout/components/TelemetryView.vue:
```diff
@@ -152,7 +152,7 @@ export default {
         unit() {
             let value = this.item.value;
-            let unit = this.metadata.value(value).unit;
+            let unit = this.meta ? this.metadata.value(value).unit : '';
```
I think this should be `this.metadata`

jvigliotta commented on src/plugins/plot/axis/XAxis.vue:
```diff
@@ -137,15 +137,17 @@ export default {
-                        key: o.key
-                    };
-                });
+                ? this.seriesModel.metadata
```
Since this is such a big block of code, can we set this.xKeyOptions to a `[]` right off the bet, then throw the actual assignment in an if block, I think that will make it more readable than a tertiary.

jvigliotta commented on src/plugins/plot/axis/YAxis.vue:
```diff
@@ -121,20 +121,22 @@ export default {
-                        key: o.key
-                    };
-                });
+                ? this.seriesModel.metadata
```
Can we do the same initial setting of `[]`, then the actual assignment in an if block based on metadata existing?

jvigliotta commented on src/plugins/plot/inspector/PlotOptionsBrowse.vue:
```diff
@@ -197,36 +197,40 @@ export default {
-                this.rangeMin = range.min;
-                this.rangeMax = range.max;
-            }
+            if (this.config) {
```
I don't know a lot about our plots, do any of these options have defaults already or need them? @shefalijoshi any thoughts on that?

alizenguyen replied:
I know plots depend a good bit on configs, so this is most likely a temp solution as well as the fix in the StackedPlotItem.vue. I'm looking into another ticket with interceptors that looks like set defaults that could be a solution? We would need to figure out what defaults to set it too though.

jvigliotta commented on src/plugins/plot/stackedPlot/StackedPlotItem.vue:
```diff
@@ -161,12 +161,17 @@ export default {
                 },
+                computed: {
+                    isMissing() {
+                        return openmct.objects.isMissing(object);
```
I think you can just pass this in as a data variable? I don't think it's state would be changed mid-use and if it was I don't think the code is set up in a way that it would trigger this computed variable update. When you assign object above, can you also just se the the isMissing value as well?

jvigliotta commented on src/plugins/plot/stackedPlot/StackedPlotItem.vue:
```diff
@@ -235,28 +240,32 @@ export default {
-                                    }
-                                ],
-                                yAxis: persistedSeriesConfig.yAxis
+                    try {
```
Would we be able to just use the `if (!this.isMissing)` here?

jvigliotta reviewed (changes requested):
Looking good! Couple suggestions and questions!

jvigliotta reviewed (approved):
LGTM!
